### PR TITLE
Fix running of Ansible playbooks shipped with tmt

### PR DESCRIPTION
**Pull request by falconizmi**
Plugins have to mangle playbook paths to appease a piece of code that
expected paths to always be relative and under the fmf root. This is not
true for playbooks shipped with plugins that are bundled with tmt, like
`prepare/feature`. Therefore changing the code to support plugins that
may never be under the fmf root.

Pull Request Checklist

* [ ] implement the feature
* [ ] write the documentation
* [ ] extend the test coverage
* [ ] update the specification
* [ ] adjust plugin docstring
* [ ] mention the version
* [ ] include a release note

### CODE REVIEW

**Comment by falconizmi** [comment]
This PR blocks #2985

File: tmt/steps/prepare/ansible.py
Comment (LecrisUT) [inline tmt/steps/prepare/ansible.py:170]: I believe this would bring it closer to addressing #2970, #2634. To that extent, I believe that we want to make the `playbook_root` point to a persistent path for all `provision` steps [^1], that way it can be reused by other guests. At the very least the inventory part should point there, but I believe roles, ansible-galaxy, and such are also valuable to be placed in a reusable path.

But I kinda see why you want to point it to `fmf_root` since you want to move the configurations from the fmf tree to `playbook_root`. Can you cook up a simple fmf tree to show how you want to reuse references from `fmf_root` in this scenario? Maybe we can think of an equivalent interface for this?

[^1]: https://github.com/teemtee/tmt/issues/2634#issuecomment-1905996432

Comment (happz) [reply tmt/steps/prepare/ansible.py:170]: It's my quick draft from yesterday which @falconizmi plans to finish, a few sentences about the motivation: fmf root gets in because of the "user-provided playbooks shall come from fmf root and nowhere else", to avoid `playbook: /usr/share/very-secret-erase-business-data.yaml` cases.

I'm not sure we could, to prevent the user from sneaking around the runner FS, replace "playbook root" with anything but the fmf root. The pre-generated, shared, and persistent inventory and host vars should be orthogonal to this problem, we still can - and should - have both at the end, plus collections and even more. Bear in mind that this PR is supposed to just do a small clean-up to unblock (correct) use of playbooks bundled with tmt & other @falconizmi's work, it probably shouldn't turn into something much bigger, but if it sparks more patches improving the Ansible situation, nobody will complain, especially not me :)

Comment (LecrisUT) [reply tmt/steps/prepare/ansible.py:170]: Ah, ok, didn't know `ansible-playbook` had a default search path for resolving playbooks, didn't quite find documentation on that. But if that's the main concern, wouldn't it be equivalent to resolve `playbook_path` to an absolute path, maybe with a check for `.is_relative_to(`/usr`)`?

But what about safeguarding against `ansible.builtin.import_playbook`? Wouldn't the user also be able to accidentally/intentionally run a protected playbook from there? Maybe setting up an `ansible.cfg` is also needed?

File: tmt/steps/prepare/feature.py
Comment (happz) [inline tmt/steps/prepare/feature.py:53]: We probably don't need this piece around, not even commented out.

File: tmt/steps/provision/__init__.py
Comment (happz) [inline tmt/steps/provision/__init__.py:1085]: TODO :)

File: tmt/steps/provision/__init__.py
Comment (happz) [inline tmt/steps/provision/__init__.py:954]: I would propose three changes:

* renaming to something like `_sanitize_playbook_path`, and
* making `playbook_root` `Optional[Path]`, and
* call it from `_run_ansible()` for all playbooks, `playbook = self._sanitize_playbook_path(playbook, playbook_root=playbook_root)`. If the root is `None`, method would return the original path, otherwise it would apply its magic.

The idea behind this is that "sanitizing" or "changing" the path argument in some way is now split between two methods: `_ansible_playbook_path()` does the work, and `_run_ansible()` decides when. That makes it harder to maintain: if we add another place that runs Ansible playbooks, we would have to duplicate this logic there as well. On the other hand, if the whole process would be handled by `_ansible_playbook_path()`, all entrypoints would simply call it and let it do the job.

File: tmt/steps/provision/__init__.py
Comment (happz) [inline tmt/steps/provision/__init__.py:959]: No need to keep no longer needed logging and asserts around.

**Comment by therazix** [comment]
It looks like this PR also fixes the #3032.

**Comment by happz** [comment]
> It looks like this PR also fixes the #3032.

Hm, yes and no. Yes, I think this PR might fix #3032, because the plugin and tmt will now use a better way when speaking about paths of playbooks that are shipped with tmt. But also no, because the cause is not the one mentioned in #3032, i.e. a missing fmf root - that's actually a *second* issue. `prepare/epel` had to perform some black magic mangling of its playbook paths, which did involve fmf root, and this PR removes this magic, therefore fmf root is no longer involved, therefore `prepare/feature` will work. But not because tmt suddenly learned to run any playbook without having an fmf root - when it comes to user-provided playbooks, like those for `prepare/ansible` or `finish/ansible`, they will still fail outside of fmf root. We discussed both issues with @falconizmi and agreed on addressing them both, in separate PRs.

**Comment by thrix** [comment]
marked as `must`, users are hitting this in Testing Farm

File: tmt/steps/prepare/feature.py
Comment (happz) [reply tmt/steps/prepare/feature.py]: Addressed in 26d311a3

File: tmt/steps/provision/__init__.py
Comment (happz) [reply tmt/steps/provision/__init__.py]: Addressed in 26d311a3

File: tmt/steps/provision/__init__.py
Comment (happz) [reply tmt/steps/provision/__init__.py]: Hopefully addressed in 26d311a3

File: tmt/steps/provision/__init__.py
Comment (happz) [reply tmt/steps/provision/__init__.py]: Addressed in 26d311a3

**Comment by happz** [comment]
Assuming @falconizmi was not available on such late notice on Friday afternoon, I tried to address the comments above. I think there will be failed tests, I'll check them in the morning.

**Comment by happz** [comment]
And before the merge, $SUBJ & commit message should be updated - it's not fixing "outside fmf root" issue, it's fixing "feature can't run its own playbooks".

**Review by lukaszachy — APPROVED** [review]

**Review by therazix — APPROVED** [review]

**Pull request merged**
